src/include/libpmemobj.h

```diff
@@ -84,6 +99,553 @@ void pmemobj_set_funcs(
+{									\
+	jmp_buf __env;							\
+	pmemobj_tx_begin(pop, &__env, locks, ##__VA_ARGS__);		\
+	setjmp(__env);							\
+	switch ((int)pmemobj_tx_get_stage()) {				\
+		case TX_STAGE_WORK:					\
+
+#define	TX_BEGIN_WRLOCK(pop, ...) __TX_BEGIN(pop, TX_LOCK_RWLOCK, ##__VA_ARGS__)
+
+#define	TX_BEGIN_LOCK(pop, ...) __TX_BEGIN(pop, TX_LOCK_MUTEX, ##__VA_ARGS__)
+
+#define	TX_BEGIN(pop) TX_BEGIN_LOCK(pop)
+
+#define	TX_ONABORT							\
+			pmemobj_tx_process();				\
+			break;						\
+		case TX_STAGE_ONABORT:
+
+#define	TX_ONCOMMIT							\
+			pmemobj_tx_process();				\
+			break;						\
+		case TX_STAGE_ONCOMMIT:
+
+/* called regardless of the transaction end state */
+#define	TX_FINALLY							\
+			pmemobj_tx_process();				\
+			break;						\
+		case TX_STAGE_FINALLY:
+
+#define	TX_END								\
+			pmemobj_tx_process();				\
+			break;						\
+		default:						\
+			pmemobj_tx_process();				\
+			break;						\
+	}								\
+	pmemobj_tx_end();						\
+}
+
+/*
+ * Takes a "snapshot" of the memory block of given size and located at given
+ * offset, and saves it in the undo log.
+ * The application is then free to directly modify the object in that memory
+ * range. In case of failure or abort, all the changes within this range will
+ * be rolled-back automatically.
+ *
+ * If successful and called during TX_STAGE_WORK, function returns zero.
+ * Otherwise, state changes to TX_STAGE_ONABORT and an error number is returned.
+ */
+int pmemobj_tx_add_range(uint64_t off, size_t size);
+
+/*
+ * Transactionally allocates a new object.
+ *
+ * If successful and called during TX_STAGE_WORK, function returns zero.
+ * Otherwise, state changes to TX_STAGE_ONABORT and an error number is returned.
+ */
+PMEMoid pmemobj_tx_alloc(size_t size, int type_num);
+
+/*
+ * Transactionally allocates new zeroed object.
+ *
+ * If successful and called during TX_STAGE_WORK, function returns zero.
+ * Otherwise, state changes to TX_STAGE_ONABORT and an error number is returned.
+ */
+PMEMoid pmemobj_tx_zalloc(size_t size, int type_num);
+
+/*
+ * Transactionally resizes an existing object.
+ *
+ * If successful and called during TX_STAGE_WORK, function returns zero.
+ * Otherwise, state changes to TX_STAGE_ONABORT and an error number is returned.
+ */
+PMEMoid pmemobj_tx_realloc(PMEMoid oid, size_t size, int type_num);
+
+/*
+ * Transactionally resizes an existing object, if extended new space is zeroed.
+ *
+ * If successful and called during TX_STAGE_WORK, function returns zero.
+ * Otherwise, state changes to TX_STAGE_ONABORT and an error number is returned.
+ */
+PMEMoid pmemobj_tx_zrealloc(PMEMoid oid, size_t size, int type_num);
+
+/*
+ * Transactionally allocates a new object with duplicate of the string s.
+ *
+ * If successful and called during TX_STAGE_WORK, function returns zero.
+ * Otherwise, state changes to TX_STAGE_ONABORT and an error number is returned.
+ */
+PMEMoid pmemobj_tx_strdup(const char *s, size_t n, int type_num);
```

Comment:
Not sure if we need that. In practice it is equivalent to: tx_alloc(strlen(s)) + memcpy.
What is the value of having a dedicated function?